File: algos/minstack/minstack.ts
// implement minstack class which has the following methods
// push, pop, top, and min element in constant time

class MinStack {
  public minElementStack: number[];
  public stack: number[];
  constructor() {
    this.minElementStack = [];
    this.stack = [];
  }

  push(el: number) {
    this.stack.push(el);
    this.minElementStack.push(
      Math.min(this.minElementStack[this.minElementStack.length - 1], el)
    );
  }
  pop() {
    this.minElementStack.pop();
    return this.stack.pop();
  }
  top() {
    return this.stack[this.stack.length - 1];
  }
  minElement() {
    return this.minElementStack[this.minElementStack.length - 1];
  }
}

File: algos/minstack/minstack.py
from typing import List
class Minstack:
  def __init__(self, stack: List[int], minElementStack: List[int]) -> None:
    self.stack = stack
    self.minElementStack = minElementStack

  def push(self, el):
    self.stack.append(el)
    self.minElementStack.append(min(el, self.minElementStack[-1]))

  def pop(self):
    self.minElementStack.pop()
    return self.stack.pop()
    
  def top(self):
    return self.stack[-1]
  
  def minElement(self):
    return self.minElementStack[-1]